### PR TITLE
Feature/signature validation

### PR DESCRIPTION
**Pull request by grimadas**
### Description of the Change

Add field validator for signatures. This validator is needed to check signatures for any Signable object.

### Benefits

Field validator can be used for any signable object 

### Possible Drawbacks 

None

### CODE REVIEW

File: shared_model/validators/field_validator.cpp [hunk 1]
```diff
@@ -228,5 +229,20 @@ namespace shared_model {
       }
     }
 
+    void FieldValidator::validateSignatures(
```
Comment (x3medima17) [inline shared_model/validators/field_validator.cpp]: Sonar sais this method is never used

File: shared_model/interfaces/common_objects/signable_hash.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ */
+
+#ifndef IROHA_SHARED_MODEL_SIGNABLE_HASH_HPP
+#define IROHA_SHARED_MODEL_SIGNABLE_HASH_HPP
+
+#include <boost/functional/hash.hpp>
+#include "interfaces/common_objects/types.hpp"
+
+namespace shared_model {
+
```
Comment (x3medima17) [inline shared_model/interfaces/common_objects/signable_hash.hpp]: Looks like extra line

File: shared_model/validators/field_validator.hpp [hunk 3]
```diff
@@ -99,6 +103,10 @@ namespace shared_model {
       void validateCounter(ReasonsGroupType &reason,
                            const interface::types::CounterType &counter) const;
 
+      void validateSignatures(ReasonsGroupType &reason,
```
Comment (x3medima17) [inline shared_model/validators/field_validator.hpp]: Why not to make `validateSignature` const and return `reason`?

Comment (kamilsa) [reply shared_model/validators/field_validator.hpp]: Then the whole API of validator should be changed . Here we modify existing reason by including new errors if any, so passing by reference makes sense

File: shared_model/backend/protobuf/block.hpp [hunk 4]
```diff
@@ -56,7 +56,7 @@ namespace shared_model {
               return HashType(proto_->payload().prev_block_hash());
             }),
             signatures_([this] {
-              SignatureSetType sigs;
+              interface::SignatureSetType sigs;
```
Comment (l4l) [inline shared_model/backend/protobuf/block.hpp]: What was the problem? I mean why did you add a namespace?

File: shared_model/interfaces/common_objects/types.hpp [hunk 5]
```diff
@@ -20,6 +20,7 @@
 
 #include <cstdint>
 #include <string>
+#include <unordered_set>
```
Comment (l4l) [inline shared_model/interfaces/common_objects/types.hpp]: Shouldn't be here

File: shared_model/validators/field_validator.cpp [hunk 6]
```diff
@@ -17,8 +17,7 @@
 
 #include "validators/field_validator.hpp"
 #include <boost/format.hpp>
-
-#include "validator/address_validator.hpp"
```
Comment (l4l) [inline shared_model/validators/field_validator.cpp]: Should be here as well:
https://github.com/hyperledger/iroha/blob/ac66a4a0801a1da551a37140fdcdc9db2243dd9e/shared_model/validators/field_validator.cpp#L88-L89

File: shared_model/validators/field_validator.cpp [hunk 7]
```diff
@@ -228,5 +229,20 @@ namespace shared_model {
       }
     }
 
+    void FieldValidator::validateSignatures(
+        ReasonsGroupType &reason,
+        const interface::SignatureSetType &signatures,
+        const crypto::Blob &source) const {
+      for (const auto &signature : signatures) {
+        if (not shared_model::crypto::CryptoVerifier<>::verify(
```
Comment (l4l) [inline shared_model/validators/field_validator.cpp]: I think you may also set alias for that in the _crypto_defaults.hpp_

File: shared_model/validators/field_validator.hpp [hunk 8]
```diff
@@ -21,7 +21,11 @@
 #include <regex>
 
 #include "datetime/time.hpp"
+#include "interfaces/base/signable.hpp"
 #include "interfaces/commands/command.hpp"
+#include "interfaces/common_objects/signable_hash.hpp"
+#include "interfaces/transaction.hpp"
+#include "validator/address_validator.hpp"
```
Comment (l4l) [inline shared_model/validators/field_validator.hpp]: Shouldn't be here

File: shared_model/validators/field_validator.hpp [hunk 9]
```diff
@@ -21,7 +21,11 @@
 #include <regex>
 
 #include "datetime/time.hpp"
+#include "interfaces/base/signable.hpp"
 #include "interfaces/commands/command.hpp"
+#include "interfaces/common_objects/signable_hash.hpp"
+#include "interfaces/transaction.hpp"
```
Comment (l4l) [inline shared_model/validators/field_validator.hpp]: `interface::Transaction` haven't been used here

File: shared_model/validators/field_validator.cpp (hunk 1 above)
Comment (grimadas) [reply shared_model/validators/field_validator.cpp]: Used now :)

File: shared_model/backend/protobuf/block.hpp (hunk 4 above)
Comment (grimadas) [reply shared_model/backend/protobuf/block.hpp]: SignatureSetType was moved

File: shared_model/validators/field_validator.cpp (hunk 6 above)
Comment (grimadas) [reply shared_model/validators/field_validator.cpp]: It is included in the header

File: shared_model/validators/signable_validator.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    template <typename ModelValidator, typename Model, typename FieldValidator>
+    class SignableModelValidator : public ModelValidator {
+      // using ModelValidator::ModelValidator;
+     public:
+      Answer validate(const Model &model) const {
+        auto answer = ModelValidator::validate(model);
+        std::string reason_name = "Signature";
+        ReasonsGroupType reason(reason_name, GroupedReasons());
+        FieldValidator().validateSignatures(
```
Comment (Warchant) [inline shared_model/validators/signable_validator.hpp]: why do you create an instance every time?
